content/blog/2019-10-08-remote-read-meets-streaming.md

```diff
@@ -0,0 +1,301 @@
+Additionally, both request and response for both methods are optionally compressed using [snappy](https://github.com/google/snappy).
+
+### Remote Write 
+
+This is the most popular way to replicate Prometheus data into 3rd party system. In this mode, Prometheus streams samples, 
+by periodically sending a batch of those to the given endpoint. 
+
+Remote read was recently improved massively in March with [WAL-based remote read](https://grafana.com/blog/2019/03/25/whats-new-in-prometheus-2.8-wal-based-remote-write/) which
+improved the reliability and resource consumption. It is also worth to note that Write is supported by almost all 3rd 
+party integrations mentioned [here](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage).
+
+### Remote Read
+
+Read method is a bit less popular. It was added in [March 2017](https://github.com/prometheus/prometheus/commit/febed48703b6f82b54b4e1927c53ab6c46257c2f) (server side) and did not change much
+since back then. Prometheus 2.13.0 release fixes the known bottlenecks of the Read, so we will focus on the Read method in this article.
+
+The key idea of the remote read is to allow querying Prometheus storage, [the TSDB](https://github.com/prometheus/prometheus/tree/master/tsdb) directly without PromQL evaluation. 
+Similar interface, [`Querier`](https://github.com/prometheus/prometheus/blob/91d7175eaac18b00e370965f3a8186cc40bf9f55/storage/interface.go#L53), PromQL engine is using to get data from storage. 
+
+The remote read request is a very simple HTTP request with the following protobuf payload:
+
+```
+message Query {
+  int64 start_timestamp_ms = 1;
+  int64 end_timestamp_ms = 2;
+  repeated prometheus.LabelMatcher matchers = 3;
+  prometheus.ReadHints hints = 4;
+}
+```
+
+With that payload, client can ask for certain series matching given `matchers` and time range with `end` and `start`.
+
+The response is equally simple. It returns the matched time series with **raw** samples of value and timestamp.
+
+```
+message Sample {
+  double value    = 1;
+  int64 timestamp = 2;
+}
+
+message TimeSeries {
+  repeated Label labels   = 1;
+  repeated Sample samples = 2;
+}
+
+message QueryResult {
+  repeated prometheus.TimeSeries timeseries = 1;
+}
+```
+
+This essentially allows read access to metrics in TSDB that Prometheus collected. The main use cases for remote read are:
+
+* Seamless Prometheus upgrades between different data formats of Prometheus, so having [Prometheus reading from another Prometheus](https://www.robustperception.io/accessing-data-from-prometheus-1-x-in-prometheus-2-0) 
+* Prometheus being able to read from 3rd party long term storage systems e.g InfluxDB.
+* 3rd party system querying data from Prometheus e.g [Thanos](https://thanos.io).
+
+## Remote Read: Problem Statement
+
+There were two key problems for such a simple remote read. It was easy to use and understand, but there were no
```

Comment:
Hmm - I don't see these as being two problems?

1) No streaming capabilities in HTTP request - why is this a problem? Were requests taking ages or something?
2) The response was including raw samples, not compressed chunks - Why is this a problem? It was harder for people to load into other prometheuses?